android/titanium/src/java/org/appcelerator/titanium/io/TiFileProvider.java

```diff
@@ -26,147 +44,402 @@ public boolean onCreate()
-		if (uriPrefix != null && uriPath.startsWith(uriPrefix)) {
-			uriPath = uriPath.substring(uriPrefix.length());
-			return new File(uriPath);
+
+		// If given content URL reference an external file, then return a File object for it.
+		String uriString = uri.toString();
+		final String FILE_SYSTEM_URI_PREFIX = getFileSystemUriPrefix();
+		if (uriString.startsWith(FILE_SYSTEM_URI_PREFIX)) {
+			Uri fileUri = Uri.parse(uriString.substring(FILE_SYSTEM_URI_PREFIX.length()));
+			return new File(fileUri.getPath());
 		}
 		return null;
 	}
 
+	/**
+	 * Converts the given file mode string to its equivalent integer file mode flags.
+	 * @param mode The file mode string to be converted such as "r" for read-only or "rw" for read/write.
+	 * @return Returns integer file mode flags matching given mode. Returns MODE_READ_ONLY if given an invalid string.
+	 */
 	private static int getFileMode(String mode)
 	{
+		int flags;
 		if ("w".equals(mode) || "wt".equals(mode)) {
-			return ParcelFileDescriptor.MODE_WRITE_ONLY | ParcelFileDescriptor.MODE_CREATE
-				| ParcelFileDescriptor.MODE_TRUNCATE;
+			flags = ParcelFileDescriptor.MODE_WRITE_ONLY;
+			flags |= ParcelFileDescriptor.MODE_CREATE;
+			flags |= ParcelFileDescriptor.MODE_TRUNCATE;
 		} else if ("wa".equals(mode)) {
-			return ParcelFileDescriptor.MODE_WRITE_ONLY | ParcelFileDescriptor.MODE_CREATE
-				| ParcelFileDescriptor.MODE_APPEND;
+			flags = ParcelFileDescriptor.MODE_WRITE_ONLY;
+			flags |= ParcelFileDescriptor.MODE_CREATE;
+			flags |= ParcelFileDescriptor.MODE_APPEND;
 		} else if ("rw".equals(mode)) {
-			return ParcelFileDescriptor.MODE_READ_WRITE | ParcelFileDescriptor.MODE_CREATE;
+			flags = ParcelFileDescriptor.MODE_READ_WRITE;
+			flags |= ParcelFileDescriptor.MODE_CREATE;
 		} else if ("rwt".equals(mode)) {
-			return ParcelFileDescriptor.MODE_READ_WRITE | ParcelFileDescriptor.MODE_CREATE
-				| ParcelFileDescriptor.MODE_TRUNCATE;
+			flags = ParcelFileDescriptor.MODE_READ_WRITE;
+			flags |= ParcelFileDescriptor.MODE_CREATE;
+			flags |= ParcelFileDescriptor.MODE_TRUNCATE;
 		} else {
-			return ParcelFileDescriptor.MODE_READ_ONLY;
+			flags = ParcelFileDescriptor.MODE_READ_ONLY;
 		}
+		return flags;
 	}
 
-	private static String getUriPrefix()
+	/**
+	 * Converts the given URI to an absolute path, if not already.
+	 * @param uri The URI to be converted. Can be null.
+	 * @return
+	 * Returns a new URI having an absolute path. Will not contain any "." or ".." in the path section.
+	 * <p>
+	 * Returns null if given a null argument.
+	 */
+	private static Uri makeAbsolute(Uri uri)
 	{
-		final TiApplication tiApp = TiApplication.getInstance();
-		if (tiApp == null) {
-			return null;
+		// Convert given URI to an absolute path in case it contains any "." or ".." segments.
+		try {
+			URI normalizedUri = URI.create(uri.toString()).normalize();
+			Uri newUri = Uri.parse(normalizedUri.toString());
+			if (newUri != null) {
+				uri = newUri;
+			}
+		} catch (Exception ex) {
```

Comment:
Maybe add a simple debug level log message in the `catch` block to avoid troubles when debugging?